cpp/include/cudf/detail/scan_reduce_iterator.cuh

```diff
@@ -0,0 +1,270 @@
+   */
+  scan_reduce_iterator(ScanIterator begin, ScanIterator end, ReduceType* last)
+    : itr_{begin}, end_{thrust::prev(end)}, last_{last}
+  {
+  }
+
+  ScanIterator itr_{};
+  ScanIterator end_{};
+  ReduceType* last_{};
+};
+
+/**
+ * @brief Create an instance of a scan_reduce_iterator
+ *
+ * @code{.pseudo}
+ *  auto begin = // begin input iterator
+ *  auto end = // end input iterator
+ *  auto reduction = rmm::device_scalar<std::size_t>(0, stream);
+ *  auto itr = make_scan_reduce_output_iterator(result,
+ *                                              result + std::distance(begin, end),
+ *                                              reduction.data());
+ *  thrust::exclusive_scan(rmm::exec_policy(stream), begin, end, output_itr, std::size_t{0});
+ *  // reduction contains the reduce result
+ * @endcode
+ *
+ * @tparam ScanIterator Output iterator type for use in a scan operation
+ * @tparam ReduceType Type used for accumulating the reduce operation
+ *
+ * @param begin Output iterator for scan
+ * @param end End of the output iterator for scan
+ * @param reduction Reduce operation result is stored here
+ * @return Instance of iterator
+ */
+template <typename ScanIterator, typename ReduceType>
+static scan_reduce_iterator<ScanIterator, ReduceType> make_scan_reduce_output_iterator(
+  ScanIterator begin, ScanIterator end, ReduceType* reduction)
+{
+  return scan_reduce_iterator<ScanIterator, ReduceType>(begin, end, reduction);
+}
+
+/**
+ * @brief Perform a combination exclusive-scan and reduce on the given input
+ *
+ * This performs an exclusive-scan and reduce (addition only) on the given input `[begin, end)`.
+ * The output of the scan is placed in `result` and the reduction result is returned.
+ *
+ * This implementation will return the reduction result in `size_t` precision regardless
+ * of the input or result types. This can be used to check if the scan will overflow
+ * when the input and result are declared as smaller types (i.e. computing offsets from sizes).
+ *
+ * Only numeric types for input and result types are supported.
+ *
+ * Note that `begin == result` is allowed but `result` may not overlap `[begin,end)` otherwise the
+ * behavior is undefined.
+ *
+ * @code{.pseudo}
+ *   auto const bytes = cudf::detail::exclusive_scan_reduce(
+ *     d_offsets, d_offsets + strings_count + 1, d_offsets, stream);
+ *   CUDF_EXPECTS(bytes <= static_cast<size_t>(std::numeric_limits<size_type>::max()),
+ *               "Size of output exceeds column size limit");
+ * @endcode
+ *
+ * @tparam ScanIterator Iterator type for input and output of the scan using addition operation
+ *
+ * @param begin Input iterator for scan/reduce
+ * @param end End of the input iterator
+ * @param result Output iterator for scan result
+ * @return Result of the reduce operation
+ */
+template <typename ScanIterator>
+std::size_t exclusive_scan_reduce(ScanIterator begin,
```

Comment:
since this looks *nearly* like a Thrust API, I suggest making it look more so by making it take an exec policy as the first parameter rather than a stream as the last. Then internally get the stream from the exec policy to pass to `reduction.value`.  This makes it more flexible for use with custom exec policies, for example.

Reply:
Would it help if I changed the name to something else? I don't expect this to become a thrust function.

Reply:
Nothing is wrong with the name. :)

Reply:
I'd prefer it look more like `cudf::detail` function than a `thrust` function.